[StaleIssueMarker] Better triage workflow

NickLaMuro opened the pull request:
There are a lot of changes here, and the commits go into greater detail, but at a high level:

- Bumps marking as `stale` to 3 months
- Auto closes pull requests that are `stale` and `unmergeable`
- Handles Pull Requests and Issues in the same manner
- When a `verified` issue becomes `stale`, remove `verified` and mark as `requires triage`
- Misc refactorings and changes to support the above


Notes
-----

- The `requires triage` label was just a label name I picked, but can be changed to whatever
  - (when the label name is decided on, that will need to be set for all repos)
- Some code was moved around to be put in `GithubService::Issue` instead of in this class.  Mostly for readability, but I could see how it might be considered a controversial change.
- The copy for marking stale and closing stale could probably use some review.
- `RSpec` `doubles` became a bit of a pain to manage in the specs, so I switch them to `GithubService::Issue` objects with valid `Sawyer::Resource` objects under the hood.  There is a little more "under the covers" knowledge about `Octokit` that is required for this, but it avoids some over stubbing that would have had to be done as a result of some of the refactoring.  I could see this also being a bit of a controversial change.
- This probably could be broken into the subcomponents described previously ( https://github.com/ManageIQ/miq_bot/issues/390#issuecomment-577715811 ), but was easier to think about this problem all at once instead of trying to address in tiny pieces.  That said, if requested I can break this up into the individual RFE items with little effort, but will leave that up to the reviewers.


Also, there is still a very big question mark around the "nudging" concept that was requested as part of #390 that I am unsure how to tackle still.


### The "bump" feature problem

The main concern here is that once the bot introduces a comment, it actually updates the `updated_at` field, and changes the search criteria pretty considerably.

To illustrate, right now we make two searches:

- All non-pinned issues/PRs that are becoming `stale`, or have been stale now for 3 months
- Any PRs that are now `stale` and `unmergeable`

The first of the above actually takes care of two of our use cases, since they both share the 3 months time window for `updated_at`, and whenever a bot interacts with a issue/PR, it resets that timer.  As a result with it, we also don't have to filter based on the `stale` label in the query, and just handle what we do with the issues that we get back from the search.


If we implement the "nudging/bump" concept, we would then get a set of search queries that look something like this:

- Un-pinned issues/PRs that have been untouched for 1 month, less than 2 months, and aren't `stale` (handle nudging)
- Un-pinned issues/PRs that have been untouched for 2 months, and aren't `stale`    (handle newly stale)
- `stale` issues/PRs that have been untouched for 3 months`                      (handle old and stale)
- Any PRs that are now `stale` and `unmergeable`                                 (handle newly unmergeable stale PRs)

Doubles our number of queries, but more importantly, it makes the whole process a bit harder to wrap my head around, and I am sure it would be harder for others as well.  Feels like there is little gain to be had for a decent amount of a logic change, but am unsure what of #390 is considered more important at this point.

Will leave the decision of pursuing the "bump" feature further to reviewers/assignies.


Links
-----

- Fixes #390 (original RFE)
- Fixes #433
- Fixes #437

NickLaMuro commented:
@miq-bot assign @mfeifer 
@miq-bot add_reviewer @Fryguy

Fryguy commented:
Why was `requires triage` added?  I don't think it's necessary.

> When an issue becomes stale and is verified, remove the [verified] label and mark it as [requires triage] to inform that triage team that this issue needs to be re-evaluated for relevancy.

Triage team will review all stale issues on a regular cadence anyway, so it would be preferable to keep the `verified` label as it was...in triage we will see it is `stale`+`verified` and make a decision on that information.

Fryguy commented on app/workers/stale_issue_marker.rb:
```diff
@@ -5,22 +5,14 @@ class StaleIssueMarker
-This issue has been automatically marked as stale because it has not been updated for at least 3 months.
+  PINNED_LABELS  = ['pinned', GithubService::Issue::REQUIRES_TRIAGE].freeze
+  SEARCH_SORTING = {:sort => :updated, :direction => :asc}.freeze
+  COMMENT_FOOTER = <<~FOOTER.sub("ManageIQ\n", "ManageIQ ").strip!
```
I don't think this is needed ... Why not just leave the multi-line comment?

NickLaMuro replied:
Cuz GithubFlavoredMarkdown...

If you leave it as multi line:

```
Thank you for all your contributions!  More information about the ManageIQ
triage process can be found in [the traige process documentation][1].

[1]: https://www.manageiq.org/docs/guides/triage_process
```

**It will format like this:**

Thank you for all your contributions!  More information about the ManageIQ
triage process can be found in [the traige process documentation][1].



**Instead of like this:**

Thank you for all your contributions!  More information about the ManageIQ triage process can be found in [the traige process documentation][1].


(make note of the line break after "ManageIQ")


[1]: https://www.manageiq.org/docs/guides/triage_process

Fryguy commented on app/workers/stale_issue_marker.rb:
```diff
@@ -5,22 +5,14 @@ class StaleIssueMarker
-
-Thank you for all your contributions!
-  EOS
+    [1]: https://github.com/ManageIQ/guides/blob/master/triage_process.md
```
Prefer https://www.manageiq.org/docs/guides/triage_process

NickLaMuro replied:
Fixed

Fryguy commented on app/workers/stale_issue_marker.rb:
```diff
@@ -5,68 +5,134 @@ class StaleIssueMarker
-    logger.info("[#{Time.now.utc}] - Closing stale PR #{pull_request.fq_repo_name}##{pull_request.number}")
-    GithubService.close_pull_request(pull_request.fq_repo_name, pull_request.number)
-    pull_request.add_comment(CLOSABLE_PR_MESSAGE)
+  def comment_copy_footer
```
empty method?

NickLaMuro replied:
Err... might have been me working out a few ideas, and then fudging it when I committed it (or a failed rebase).

Will look into it.

NickLaMuro replied:
Fixed

Fryguy commented:
Overall LGTM. aside from the removal of the "requires triage" label as mentioned above.

Fryguy commented:
> RSpec doubles became a bit of a pain to manage in the specs, so I switch them to GithubService::Issue objects with valid Sawyer::Resource objects under the hood. There is a little more "under the covers" knowledge about Octokit that is required for this, but it avoids some over stubbing that would have had to be done as a result of some of the refactoring. I could see this also being a bit of a controversial change.

Can you explain this a bit more...I saw in your other PRs you were going in this direction, and I don't understand why the need for something that appears more complicated at a glance.

Fryguy commented:
> Also, there is still a very big question mark around the "nudging" concept that was requested as part of #390 that I am unsure how to tackle still.

We can do that in a follow up I think.

NickLaMuro commented:
@Fryguy This is specifically what I was asking you for clarification on here:

https://github.com/ManageIQ/miq_bot/issues/390#issuecomment-591657256

Specifically around this part where I tried to interpret:

> "If a bug is verified, and reaches the stale time period, remove the verified label so that it can be re-triaged, but don't mark as stale."

Which, the `requires triage` was meant to be a signal that recently both were removed (per the requirements from that issue).  Also:

> in triage we will see it is stale+verified and make a decision on that information.

This is the opposite of what was asked of me to do here:

https://github.com/ManageIQ/miq_bot/issues/390#issuecomment-577715811


* * *

Please clarify the requirements before I try and argue this further, as I don't want to be coding something up where there isn't even an agreement on how this should function.

Fryguy commented:
Ah! thanks for the context...I never saw that comment hence why I was so confused.  Now I understand why you added the `requires triage`.

When we wrote up those rules it was not with a bot in mind but with a human process...that is, when we in the triage meeting saw `stale` + `verified` we would *choose* to remove verified if it needed re-verification.  But that is an action only humans can take (IMO), but that didn't really get reflected in the requirements.

The wording does not reflect that though, so that's definitely something we should change in the triage_process.md .

I would say the bot should just add the stale label, which we will see in triage and we handle what to do with the verified label at that time.

NickLaMuro commented:
> Can you explain this a bit more...

Sure.

Previously, this was one of the example stubs:

```ruby
double("stale_pr",
       :updated_at    => 7.months.ago,
       :fq_repo_name  => fq_repo_name,
       :number        => 2,
       :pull_request? => true,
       :labels        => [])
```

Specifically, the `labels` was stubbing the call to `.labels` directly, so that worked fine, but now that we have moved to using `GithubService::Issue#stale?` and `GithubService::Issue#.unmergeable?`, that no longer works since the `double` try and implement the logic for `#stale?` and `#unmergeable?` and use `:labels`.

So I had two options:

- Change the double to instead stub `#stale?` and `#unmergeable?` as booleans
- Do what I did here which instead effectively does the "`FactoryBot.build`" approach

I chose the latter route since it ends up exercising more of our codebase instead of hiding it behind stubs.

Unit test purists probably hate it since it is now testing outside the scope of this model, but I brought up `FactoryBot` (or `ActiveRecord::Base#new`) since it is pretty common practice to break that rule.  And since we only started testing any part of `GithubService` object (minus the `github_service/commands/) when I added the spec for it here:

https://github.com/ManageIQ/miq_bot/pull/481/files#diff-74e22cb2c056bfcfc22ba9006730335c

I figured it is probably better to actually test as far down the stack as possible instead of stubbing it away.

Also, since I am aware a rework of octokit is in the works:

https://github.com/octokit/octokit.rb/pull/1199#issuecomment-589970552

I think testing with as much of the internals of that gem is ideal so that when we inevitably upgrade that gem to the new release.  If we stub everything out, we don't catch anything with tests. 

> I saw in your other PRs you were going in this direction, and I don't understand why the need for something that appears more complicated at a glance.

This is different than probably what you are referring to, which is was the [`github_service_helper.rb`](https://github.com/ManageIQ/miq_bot/blob/4325a17/spec/support/github_service_helper.rb#L1) and that is not what I am using here.  Though, I probably could have, but decided to leave the existing tests as close to as they were as possible.

In this case, the specs already stubbed the fetch to github and returned the dummy objects from the stub.  If I were to have used `github_service_helper.rb`, the stub/dummy objects wouldn't have been necessary and only creating a stub response would have been required.

Functionally the amount of code that gets exercised is similar to what I did here, but two different approaches to doing that.  Again, was trying to reduce the diff of the original specs here to maintain their validity, so I chose not to do that as a result.

Fryguy commented:
Thanks for the info...yeah I was talking more about the github_service_helper, but also was curious about the change approach in general.
